packages/loader/container-loader/src/serializedStateManager.ts

```diff
@@ -119,10 +140,93 @@ export class SerializedStateManager {
 		} else {
 			const { baseSnapshot, snapshotBlobs } = this.pendingLocalState;
 			this.snapshot = { baseSnapshot, snapshotBlobs };
+			this.refreshSnapshot(supportGetSnapshotApi).catch((error) => {
+				this.mc.logger.sendErrorEvent(
+					{
+						eventName: "Could_Not_Refresh_Snapshot",
+					},
+					error,
+				);
+				throw error;
+			});
 			return { baseSnapshot, version: undefined };
 		}
 	}
 
+	/**
+	 * Fetch latest snapshot available in storage and retry indefinetely
+	 * @param supportGetSnapshotApi -
+	 */
+	private async refreshSnapshot(supportGetSnapshotApi: boolean) {
+		// This need to be resilient to failure. We can use runWithRetry
+		const { snapshot } = supportGetSnapshotApi
+			? await fetchISnapshot(this.mc, this.storageAdapter, undefined)
+			: await fetchISnapshotTree(this.mc, this.storageAdapter, undefined);
+		const latestSnapshot: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
+			? snapshot.snapshotTree
+			: snapshot;
+		assert(latestSnapshot !== undefined, 0x8e4 /* Snapshot should exist */);
```

Comment:
this code is duplicated a few places. maybe integrate with or create free function that handles this,.

also, we shouldn't duplicateassert short codes. they should all be unique. however, if you move this all to a shared place, so there is only one, copying the existing code there is fine